File: 백준/Gold/1062. 가르침/가르침.cc
#include <bits/stdc++.h>

#define MAX 50

using namespace std;

int N, K;
int Max;
int word[MAX];
int checked;

void DFS(int pick, int start, int check) {
	if (pick == 0) {
		int cnt = 0;
		for (int i = 0; i < N; i++) {
			if ((word[i] & check) == word[i]) cnt++;
		}
		if (Max < cnt) Max = cnt;
		return;
	}

	for (int i = start; i < 26; i++) {
		if ((check & (1 << i)) == 0) {
			check |= (1 << i);
			DFS(pick - 1, i, check);
			check &= ~(1 << i);
		}
	}
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);
	cout.tie(nullptr);

	cin >> N >> K;
	string str;
	for (int i = 0; i < N; i++) {
		cin >> str;
		int num = 0;
		for (char c: str) {
			num |= 1 << (c - 'a');
		}
		word[i] = num;
	}

	if (K < 5) {
		printf("0");
		return 0;
	}
	if (K == 26) {
		printf("%d", N);
		return 0;
	}
	char antic[] = "antic";
	for (char c: antic) {
		checked |= 1 << (c - 'a');
	}
	DFS(K - 5, 0, checked);
	printf("%d", Max);


	return 0;
}

File: 백준/Gold/1062. 가르침/가르침.js
const fs = require('fs');
const file = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = fs.readFileSync(file).toString().trim().split('\n');

const [, K] = input[0].split(' ').map(Number);

const words = input.slice(1).map((str) => str.trim());

function solution(words, k) {
  if (k < 5) {
    console.log(0);
    return;
  }

  if (k === 26) {
    console.log(words.length);
    return;
  }

  const bitWords = words.map((word) => convertToBit(word));
  const checked = convertToBit('antic');
  let max = 0;

  const find = (pick, start, checked) => {
    if (pick === 0) {
      const count = bitWords.reduce((sum, word) => ((word & checked) === word ? sum + 1 : sum), 0);
      max = Math.max(max, count);
      return;
    }

    for (let i = start; i < 26; i++) {
      if ((checked & (1 << i)) !== 0) continue;

      checked |= 1 << i;
      find(pick - 1, i, checked);
      checked &= ~(1 << i);
    }
  };

  find(k - 5, 0, checked);

  console.log(max);
}

const convertToBit = (word) =>
  word.split('').reduce((sum, current) => sum | (1 << (current.charCodeAt(0) - 97)), 0);

solution(words, K);
